File: Buoi2/C++/Quick Sort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Hàm Partition để phân chia mảng dựa trên pivot
int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];  // Chọn pivot là phần tử cuối
    int i = low - 1;  // Chỉ số nhỏ hơn pivot
    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);  // Hoán đổi arr[i] và arr[j]
        }
    }
    swap(arr[i + 1], arr[high]);  // Hoán đổi pivot vào vị trí đúng
    return i + 1;
}

// Hàm Quick Sort đệ quy
void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int p = partition(arr, low, high);  // Tìm vị trí pivot
        quickSort(arr, low, p - 1);  // Sắp xếp phần bên trái của pivot
        quickSort(arr, p + 1, high);  // Sắp xếp phần bên phải của pivot
    }
}

int main() {
    vector<int> arr = {5, 2, 9, 1, 5, 6};
    quickSort(arr, 0, arr.size() - 1);
    cout << "Kết quả Quick Sort: ";
    for (int x : arr) {
        cout << x << " ";
    }
    cout << endl;
    return 0;
}

File: Buoi2/Julia/Quick Sort.jl
# Hàm Partition để phân chia mảng dựa trên pivot
function partition!(arr, low, high)
    pivot = arr[high]  # Chọn pivot là phần tử cuối
    i = low - 1  # Chỉ số nhỏ hơn pivot
    for j in low:high-1
        if arr[j] <= pivot
            i += 1
            arr[i], arr[j] = arr[j], arr[i]  # Hoán đổi arr[i] và arr[j]
        end
    end
    arr[i + 1], arr[high] = arr[high], arr[i + 1]  # Hoán đổi pivot vào vị trí đúng
    return i + 1
end

# Hàm Quick Sort đệ quy
function quick_sort!(arr, low, high)
    if low < high
        p = partition!(arr, low, high)  # Tìm vị trí pivot
        quick_sort!(arr, low, p - 1)  # Sắp xếp phần bên trái của pivot
        quick_sort!(arr, p + 1, high)  # Sắp xếp phần bên phải của pivot
    end
end

arr = [5, 2, 9, 1, 5, 6]
quick_sort!(arr, 1, length(arr))
println("Kết quả Quick Sort: ", arr)
